langchain/src/llms/watsonx_ai.ts

```diff
@@ -0,0 +1,180 @@
+   */
+  ibmCloudApiKey?: string;
+  /**
+   * WatsonX AI Key.
+   * Provide API Key if you do not wish to automatically pull from env.
+   */
+  projectId?: string;
+  /**
+   * Parameters accepted by the WatsonX AI Endpoint.
+   */
+  modelParameters?: Record<string, unknown>;
+  /**
+   * WatsonX AI Model ID.
+   */
+  modelId?: string;
+}
+
+const endpointConstructor = (region: string, version: string) =>
+  `https://${region}.ml.cloud.ibm.com/ml/v1-beta/generation/text?version=${version}`;
+
+/**
+ * The WatsonxAI class is used to interact with Watsonx AI
+ * Inference Endpoint models. It uses IBM Cloud for authentication.
+ * This requires your IBM Cloud API Key which is autoloaded if not specified.
+ */
+
+export class WatsonxAI extends LLM<BaseLLMCallOptions> {
+  static lc_name() {
+    return "WatsonxAI";
+  }
+
+  get lc_secrets(): { [key: string]: string } | undefined {
+    return {
+      ibmCloudApiKey: "IBM_CLOUD_API_KEY",
+      projectId: "WATSONX_PROJECT_ID",
+    };
+  }
+
+  endpoint: string;
+
+  region = "us-south";
+
+  version = "2023-05-29";
+
+  modelId = "meta-llama/llama-2-70b-chat";
+
+  modelKwargs?: Record<string, unknown>;
+
+  ibmCloudApiKey?: string;
+
+  projectId?: string;
+
+  modelParameters?: Record<string, unknown>;
+
+  constructor(fields: WatsonxAIParams) {
+    super(fields);
+
+    this.region = fields?.region ?? this.region;
+    this.version = fields?.version ?? this.version;
+    this.modelId = fields?.modelId ?? this.modelId;
+    this.ibmCloudApiKey =
+      fields?.ibmCloudApiKey ?? getEnvironmentVariable("IBM_CLOUD_API_KEY");
+    this.projectId =
+      fields?.projectId ?? getEnvironmentVariable("WATSONX_PROJECT_ID");
+
+    this.endpoint =
+      fields?.endpoint ?? endpointConstructor(this.region, this.version);
+    this.modelParameters = fields.modelParameters;
+
+    if (!this.ibmCloudApiKey) {
+      throw new Error("Missing IBM Cloud API Key");
+    }
+
+    if (!this.projectId) {
+      throw new Error("Missing WatsonX AI Project ID");
+    }
+  }
+
+  _llmType() {
+    return "WatsonxAI";
+  }
+
+  /**
+   * Calls the WatsonX AI endpoint and retrieves the result.
+   * @param {string} prompt The input prompt.
+   * @returns {Promise<string>} A promise that resolves to the generated string.
+   */
+  /** @ignore */
+  async _call(
+    prompt: string,
+    _options: this["ParsedCallOptions"]
+  ): Promise<string> {
+    interface WatsonxAIResponse {
+      results: {
+        generated_text: string;
+        generated_token_count: number;
+        input_token_count: number;
+      }[];
+      errors: {
+        code: string;
+        message: string;
+      }[];
+    }
+    const response = (await this.caller.call(async () =>
+      fetch(this.endpoint, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          Accept: "application/json",
+          Authorization: `Bearer ${await this.generateToken()}`,
```

Comment:
Should we be caching this or is it expected to generate a new access token on each model call?

Reply:
That is a great point. The token is reusable for up to an hour by default, so likely makes sense to cache.
I have updated `generateToken` to reuse the token if it is still valid.